pkg/test/syncset/syncset.go

```diff
@@ -0,0 +1,94 @@
+package syncset
+
+import (
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+// Option defines a function signature for any function that wants to be passed into BuildSyncSet
+type Option func(*hivev1.SyncSet)
+
+// Checksum defines a function signature for checksumming objects.
+type Checksum func(*hivev1.SyncSet) (string, error)
+
+// BuildSyncSet runs each of the functions passed in to generate a cluster deployment.
+func BuildSyncSet(opts ...Option) *hivev1.SyncSet {
+	retval := &hivev1.SyncSet{}
+	for _, o := range opts {
+		o(retval)
+	}
+
+	return retval
+}
+
+// WithName sets the SyncSet.Name field when building a syncset with BuildSyncSet.
+func WithName(name string) Option {
+	return func(syncSet *hivev1.SyncSet) {
+		syncSet.Name = name
+	}
+}
+
+// WithNamePostfix appends the string passed in to the SyncSet.Name field when building a SyncSet with BuildSyncSet.
+func WithNamePostfix(postfix string) Option {
+	return func(syncSet *hivev1.SyncSet) {
+		syncSet.Name = syncSet.Name + "-" + postfix
+	}
+}
+
+// WithNamespace sets the SyncSet.Namespace field when building a SyncSet with BuildSyncSet.
+func WithNamespace(namespace string) Option {
+	return func(syncSet *hivev1.SyncSet) {
+		syncSet.Namespace = namespace
+	}
+}
+
+// WithAnnotationsSet ensures that SyncSet.Annotations is not nil.
+func WithAnnotationsSet() Option {
+	return func(syncSet *hivev1.SyncSet) {
+		// Only set if Nil (don't wipe out existing)
+		if syncSet.Annotations == nil {
+			syncSet.Annotations = map[string]string{}
+		}
+	}
+}
+
+// WithEmptyAnnotations ensures that annotations is set and that map is empty.
+func WithEmptyAnnotations() Option {
+	return func(syncSet *hivev1.SyncSet) {
+		// Wipe out existing so that it's actually empty.
+		syncSet.Annotations = map[string]string{}
+	}
+}
+
+// WithEmptyBackupChecksum ensures that the checksum annotation is empty.
+func WithEmptyBackupChecksum() Option {
+	return func(syncSet *hivev1.SyncSet) {
+		// Case where annotations don't exist, so checksum doesn't exist.
+		if syncSet.ObjectMeta.Annotations == nil {
+			return
+		}
+
+		// Case where Annotations exist, -and- checksum annotation exists.
+		delete(syncSet.Annotations, controllerutils.LastBackupAnnotation)
+
+		// If that was the only entry, then we should nil out annotations (this makes comparisons with FakeClient work).
+		if len(syncSet.Annotations) == 0 {
+			syncSet.Annotations = nil
+		}
+	}
+}
+
+// WithBackupChecksum sets the object's checksum attribute to the correct value.
+func WithBackupChecksum(checksum Checksum) Option {
+	return func(syncSet *hivev1.SyncSet) {
+		// In order for the checksum to be correct, the checksum needs to be calculated AFTER all other changes have been made
```

Comment:
minor nit. i'd prefer if this comment was up with the function doc three lines up (so that i can see this in VSCode when calling the func :) )